woocommerce/woocommerce-gutenberg-products-block
# Product Gallery block: Restrict block to be available only on the Single Product template or the Product Gallery template part

## What
- Add a TemplateChangeDetector that detects when a template or page changes and notifies any subscribers (currently only the BlockRegistrationManager).
- Add the BlockRegistrationManager, which is responsible for registering/unregistering blocks according to the current template/page.
- In a subsequent PR, we can replace the current implementation of registerBlockSingleProductTemplate with this new pattern, which is more performant and less prone to racing conditions that have already caused issues in the past.

Fixes #11027

## Why
The current implementation of the Product Gallery block does not work outside the Single Product template context (or the Product Gallery template part, which is responsible for the dialog when clicking on the Large Image block). Therefore, we need to restrict the block to be available only within this context.



## Testing Instructions



_Please consider any edge cases this change may have, and also other areas of the product this may impact._

#### Test 1: Single Product template
1. From your WordPress dashboard, go to Appearance > Themes. Select a block-based theme like "Twenty-twenty Three," "Twenty-twenty Four," etc.
2. On the left-hand side menu, click on Appearance > Editor > Templates
3. Find and select the 'Single Product' template from the list.
4. When the Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before.
5. Inside the Page editor, click on the '+' button to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block. Click on it to add the block to the template.
7. Make sure the Product Gallery block can be added to the editor.
8. Click on Save.
9. Visit a product page of a Variable Product and make sure the block is working as expected.

#### Test 2: Posts
1. From your WordPress dashboard, go to Appearance > Themes. Select a block-based theme like "Twenty-twenty Three," "Twenty-twenty Four," etc.
2. On the left-hand side menu, click on Posts > Add new.
5. Inside the Page editor, click on the '+' button to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block.
7. Make sure the Product Gallery block DOES NOT appear and cannot be added to the editor.

#### Test 3: Pages
1. From your WordPress dashboard, go to Appearance > Themes. Select a block-based theme like "Twenty-twenty Three," "Twenty-twenty Four," etc.
2. On the left-hand side menu, click on Pages > Add new.
5. Inside the Page editor, click on the '+' button to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block.
7. Make sure the Product Gallery block DOES NOT appear and cannot be added to the editor.


## WooCommerce Visibility


Required:
